### PR TITLE
Add new auto scale feature match method based on clustering.

### PR DESCRIPTION
**Pull request by wujunkai166**

### CODE REVIEW

**Comment by wujunkai166** [comment]
@windyuan

File: modules/ocl/cl_geo_map_handler.h [hunk 1]
```diff
@@ -55,6 +55,7 @@ class GeoKernelParamCallback
     virtual PointFloat2 get_left_scale_factor () = 0;
     virtual PointFloat2 get_right_scale_factor () = 0;
 
+    virtual const BowlDataConfig &get_bowl_config () = 0;
```
Comment (windyuan) [inline modules/ocl/cl_geo_map_handler.h]: It doesn't make sense to add get_bowl_config.
GeoMapHandler should not be aware of BowlConfig. if have to add a interface. maybe try 
virtual float get_stable_y_ratio();
The ratio must have a default value for other generic cases.

this ratio can be set in 360_stitching or fisheye.

File: modules/ocl/cl_geo_map_handler.h [hunk 2]
```diff
@@ -135,6 +136,13 @@ class CLGeoMapHandler
         return PointFloat2 (0.0f, 0.0f);
     }
 
+    virtual const BowlDataConfig &get_bowl_config () {
+        XCAM_ASSERT (false && "CLGeoMapHandler::get_bowl_config is not supported");
+        BowlDataConfig bowl_data_config;
```
Comment (windyuan) [inline modules/ocl/cl_geo_map_handler.h]: This function should be rejected.
BTW, this looks like hardcode and usage of variable reference is wrong. C++reference(&) is similar like pointer.  bowl_data_config is a local variable, after function returned, the reference is damaged. the return value can be any values. BE AWARE OF THAT. if your result is not correct, it's much possible caused by this function.

File: cl_kernel/kernel_geo_map.cl [hunk 3]
```diff
@@ -67,6 +67,7 @@ kernel_geo_map (
     __read_only image2d_t input_y, __read_only image2d_t input_uv,
     __read_only image2d_t geo_table, float2 table_scale_size,
     float2 left_scale_factor, float2 right_scale_factor,
+    float wall_image_height,
```
Comment (windyuan) [inline cl_kernel/kernel_geo_map.cl]: change wall_image_height to other name maybe stable_height or smthing.

**Comment by wujunkai166** [comment]
@windyuan  
Changes compared to last patch:
1. Class CVFeatureMatchCluster inherited from CVFeatureMatch instead of FeatureMatch.
2. Move out get_bowl_config() in geo map handler with get_stable_y_start() replaced.
3. Add member variable _stable_y_start in fisheye handler and initialize it to 0.0f when it is in general cases other than "BowlView".

File: modules/ocl/cv_feature_match_cluster.cpp [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+            if (valid_seeds[i]) {
+                seed_x_offset = corner1[i].x - corner0[i].x;
+                seed_y_offset = corner1[i].y - corner0[i].y;
+                cluster.push_back (i);
+                cluster_offset.push_back (cv::Point2f(seed_x_offset, seed_y_offset));
+                valid_corners[i] = 0;
+                valid_seeds[i] = 0;
+                break;
+            }
+        }
```
Comment (windyuan) [inline modules/ocl/cv_feature_match_cluster.cpp]: is it possible to cause a dead loop here while(1). and something like kwscan may report issue either.

Comment (wujunkai166) [reply modules/ocl/cv_feature_match_cluster.cpp]: It won't cause a dead loop but I'll modify the condition to be explicit anyway.

**Comment by wujunkai166** [comment]
@windyuan

**Comment by windyuan** [comment]
merged. thanks.

**Pull request merged**